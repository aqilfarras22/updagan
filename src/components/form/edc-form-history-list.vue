<template>
  <table-component
    showEmpty
    noGutters
    :items="history"
    :fields="formHistoryFields"
  >
    <!-- date -->
    <template slot="createdAt" slot-scope="data">
      {{ new Date(data.row.item.createdAt) | formatUTCDate }}
    </template>
    <!-- action -->
    <template slot="eventType" slot-scope="data">
      {{ data.row.item.primaryReason && data.row.item.primaryReason!='' ? data.row.item.primaryReason : (verificationStatusText[data.row.item.eventType] || approvalStatusText[data.row.item.eventType] || historyStatusText[data.row.item.eventType]) }}
    </template>
    <!-- view -->
    <template slot="view" slot-scope="data">
      <a v-if="historyStatus.MODIFIED === data.row.item.eventType" @click.prevent="view(data.row.item)" href="">View</a>
    </template>
  </table-component>
</template>

<script>
import TableComponent from '@/components/common/table';
import {
  approvalStatusText,
  verificationStatusText,
  historyStatusText,
  historyStatus,
} from '@/constants/form';

export default {
  name: 'form-history-list',
  components: {
    TableComponent,
  },
  props: {
    history: {
      type: Array,
      default: null,
    },
  },
  data() {
    return {
      approvalStatusText,
      verificationStatusText,
      historyStatusText,
      historyStatus,
      formHistoryFields: [
        {
          key: 'createdAt',
          label: 'Date',
          sortable: true,
        },
        {
          key: 'eventType',
          label: 'Action / Changes',
          sortable: true,
        },
        {
          key: 'createdBy',
          label: 'By',
          sortable: true,
        },
        {
          key: 'view',
          label: ' ',
        },
      ],
    };
  },
  methods: {
    view(item) {
      this.$emit('view', item);
    },
  },
};
</script>
