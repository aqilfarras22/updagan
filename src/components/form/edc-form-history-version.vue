<template>
  <div>
    <b-row class="border-bottom header-row">
      <b-col><a @click.prevent="back" class="backButton" href=""><i class="fa fa-caret-left fa-lg"></i>&nbsp; Back to history</a></b-col>
      <b-col class="text-right"><i class="square"></i>&nbsp; <span class="info">Edited fields</span></b-col>
    </b-row>
    <div>
      <b-card-group v-if="!edcFormVersionData.isUnSchedule" class="pb-15">
        <b-card no-body class="border-0" :header="edcFormVersionData.name">
          <div v-if="edcFormVersionData.primaryReason" class="reason">
            Reason for change: {{ edcFormVersionData.primaryReason }}
          </div>
          <SubmittedFormPreview v-if="edcFormVersionData.fields" :isHistory="true" :formData="edcFormVersionData.fields" />
        </b-card>
      </b-card-group>
    </div>
    <div v-if="edcFormVersionData.isUnSchedule && edcFormVersionData.fields">
      <b-card-group v-for="(unscheduleAnswers, index) in edcFormVersionData.fields" :key="index" :class="{'border-top': index!=0 ,'pb-15' : index+1 === edcFormVersionData.fields.length, 'pb-5' : index+1 !== edcFormVersionData.fields.length}">
        <b-card no-body class="border-0" :header="unscheduleAnswers.formName">
          <div v-if="edcFormVersionData.primaryReason" class="reason">
            Reason for change: {{ edcFormVersionData.primaryReason }}
          </div>
          <SubmittedFormPreview v-if="unscheduleAnswers.formData" :isHistory="true" :formData="unscheduleAnswers.formData"/>
        </b-card>
      </b-card-group>
    </div>
  </div>
</template>

<script>
import SubmittedFormPreview from '@/components/edc-form/submitted-form-preview';

export default {
  name: 'form-history-version',
  components: { SubmittedFormPreview },
  props: {
    edcFormVersionData: Object,
  },
  methods: {
    back() {
      this.$emit('back');
    },
  },
};
</script>

<style scoped>
.header-row {
  padding: 0.9rem 0 0.65rem;
}
.square {
  display: inline-block;
  width: 16px;
  height: 16px;
  border-radius: 1px;
  margin-bottom: -2px;
  background-color: #ffa200;
}
.info {
  font-size: 12px;
  display: inline-block;
}
.reason {
  font-family: "RobotoBold" !important;
  color: #666669;
  padding-left: 30px;
  padding-top: 19px;
}
</style>
