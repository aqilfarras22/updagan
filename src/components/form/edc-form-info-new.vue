<template>
  <div class="edc-form-bar">
	<div class="row">
		<div class="col-md-2">
			<div class="heading">Form/Packet Name</div>
			<div class="value">{{ clearFormName() }}</div>
		</div>
		<div class="col-md-2">
			<div class="heading">Verification Status</div>
			<div class="value" :class="edcFormInfo.verificationStatus">{{ verificationStatusText[edcFormInfo.verificationStatus] }}</div>
			<div v-if="verified && edcFormInfo.verifiedOn" class="date value">
				({{ new Date(edcFormInfo.verifiedOn) | formatGMTDate }})
			</div>
		</div>
		<div class="col-md-2">
			<div class="heading">Approval Status</div>
			<div class="value" :class="edcFormInfo.approvalStatus">{{ approvalStatusText[edcFormInfo.approvalStatus] }}</div>
			<span v-if="approved && edcFormInfo.approvedOn" class="date value">
			({{ new Date(edcFormInfo.approvedOn) | formatGMTDate }})
			</span>
		</div>
	</div>
  </div>
</template>

<script>
import { AclRule } from 'vue-acl';
import { mapGetters } from 'vuex';
import {
  approvalStatusText,
  verificationStatusText,
  verificationStatus,
  approvalStatus,
} from '@/constants/form';

export default {
  name: 'form-info',
  components: {},
  props: {
    edcFormInfo: {
		type: Object,
		default: {}
	}
  },
  data() {
    return {
      approvalStatusText,
	  verificationStatusText,
	  verificationStatus
	};
  },
  computed: {
    ...mapGetters(['user']),
    userAccess() {
      return new AclRule(this.user.userRole).generate();
    }
  },
  methods: {
	clearFormName() {
    	return this.edcFormInfo.formName ? this.edcFormInfo.formName.replace(/_/g, ' ') : '';
	},
	verified() {
      	return this.edcFormInfo.verificationStatus === verificationStatus.VERIFIED;
    },
    approved() {
      	return this.edcFormInfo.approvalStatus === approvalStatus.APPROVED;
    }
  }
};
</script>

<style scoped lang="scss">
.edc-form-bar {
	height: 60px;
  background-color: #fff;
  padding: 10px 0px;
  -webkit-box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.06);
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.06);
  margin-top: 15px;
  height: 75px;
  .heading {
    font-weight: 600;
  }
}
.date {
  font-size: 12px;
  line-height: 1.67;
}
.telehealth-call {
  .heading {
    font-size: 10px;
  }
  .value {
    font-size: 10px;
  }
}
</style>
