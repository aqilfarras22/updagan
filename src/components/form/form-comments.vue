<template>
  <div>
    <b-card-group class="border-top">
      <b-card no-body class="border-0" :header="title">
        <b-list-group v-if="comments" flush>
          <b-list-group-item class="item border-0" v-for="comment in comments" :key="comment.id">
            <b-row>
              <div class="col-5">
                <div class="name-class pb-5">
                  <span v-if="comment.author.name.firstName">{{ comment.author.name.firstName }}</span>
                  <span v-if="comment.author.name.middleName">{{ comment.author.name.middleName }}</span>
                  <span v-if="comment.author.name.lastName">{{ comment.author.name.lastName }}</span>
                </div>
                <div>{{ new Date(comment.postedAt * 1000 ) | formatGMTDate }}</div>
              </div>
              <div class="col-7 div_style">{{ comment.message }}</div>
            </b-row>
          </b-list-group-item>
        </b-list-group>
      </b-card>
    </b-card-group>
  </div>
</template>

<script>
import ButtonCmp from '@/components/common/button';

export default {
  name: 'form-fields-list',
  components: { ButtonCmp },
  props: {
    comments: Array,
  },
  data() {
    return {
      title: 'Comments',
    };
  },
  methods: {},
};
</script>

<style scoped>
.item {
  font-size: 0.9rem;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.43;
  text-align: left;
  color: #666669;
  padding-left: 30px;
}
.name-class {
  font-weight: 700;
}
.div_style {
  border-left: 1px solid #d4d9e2; 
}
</style>
