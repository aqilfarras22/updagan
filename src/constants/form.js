export const fieldTypes = {
  POLAR_QUESTION: 'POLAR_QUESTION',
  DROPDOWN: 'DROPDOWN',
  DATEPICKER: 'DATEPICKER',
  INPUT: 'INPUT',
  LABEL: 'LABEL',
};

export const formFieldQueryStatus = {
  OPEN: 'OPEN',
  CLOSE: 'CLOSE',
};

export const historyStatus = {
  MODIFIED: 'MODIFIED',
};

export const historyStatusText = {
  MODIFIED: 'Original Submission',
};

// approve
export const approvalStatus = {
  APPROVED: 'APPROVED',
  NOT_APPROVED: 'NOT_APPROVED',
};
export const approvalStatusText = {
  APPROVED: 'Approved',
  NOT_APPROVED: 'Unapproved',
};

// verification
export const verificationStatus = {
  VERIFIED: 'VERIFIED',
  NOT_VERIFIED: 'NOT_VERIFIED',
};
export const verificationStatusText = {
  VERIFIED: 'Verified',
  NOT_VERIFIED: 'Unverified',
};
