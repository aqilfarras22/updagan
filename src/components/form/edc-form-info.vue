<template>
  <div v-if="formId">
    <dl v-if="formName" ref="formName">
      <dt>Form/Packet Name</dt>
      <dd>{{ clearFormName }}</dd>
    </dl>
    <dl ref="verificationStatus">
      <dt>Verification Status</dt>
      <dd>
        <div :class="verificationStatus">{{ verificationStatusText[verificationStatus] }}</div>
        <div v-if="verified && verifiedOn" class="date">
          ({{ new Date(verifiedOn) | formatGMTDate }})
        </div>
      </dd>
    </dl>
    <dl ref="approvalStatus">
      <dt>Approval Status</dt>
      <dd>
        <div :class="approvalStatus">{{ approvalStatusText[approvalStatus] }}</div>
        <span v-if="approved && approvedOn" class="date">
          ({{ new Date(approvedOn) | formatGMTDate }})
        </span>
      </dd>
    </dl>
    <div class="actions">
      <button-cmp v-if="canPiAccess || canSubPiAccess || canScAccess || canHomeHealthAccess" class="form-action-button" size="lg" variant="primary" @click="openApprovalModal">{{ approvalBtnText }}</button-cmp>
      <button-cmp v-if="canCraAccess" class="form-action-button" size="lg" variant="primary" @click="openVerificationModal">{{ verifyBtnText }}</button-cmp>
    </div>
    <formApprovalModal
      v-if="canPiAccess || canSubPiAccess || canScAccess || canHomeHealthAccess"
      ref="approvalModal"
      callApi
      :formId="formId"
      :approved="approved"
      @confirmed="approvalConfirmed"
    />
    <formVerificationModal
      v-if="canCraAccess"
      ref="verificationModal"
      callApi
      :formId="formId"
      :verified="verified"
      @confirmed="verificationConfirmed"
    />
  </div>
</template>

<script>
import { AclRule } from 'vue-acl';
import { mapGetters } from 'vuex';
import formVerificationModal from '@/components/modals/edc-form-verification-modal';
import formApprovalModal from '@/components/modals/edc-form-approval-modal';
import ButtonCmp from '@/components/common/button';

import { PI, CRA, SUB_PI, STUDY_COORDINATOR, HOME_HEALTH } from '@/constants/authRoles';
import {
  approvalStatusText,
  verificationStatusText,
  verificationStatus,
  approvalStatus,
} from '@/constants/form';

export default {
  name: 'form-info',
  components: {
    formVerificationModal,
    formApprovalModal,
    ButtonCmp,
  },
  props: {
    formId: String,
    formName: String,
    approvalStatus: String,
    verificationStatus: String,
    verifiedOn: [String, Number],
    approvedOn: [String, Number],
  },
  data() {
    return {
      approvalStatusText,
      verificationStatusText,
    };
  },
  computed: {
    ...mapGetters(['user']),
    userAccess() {
      return new AclRule(this.user.userRole).generate();
    },
    canPiAccess() {
      this.$acl.change(`${PI}`);
      return this.$acl.check('userAccess');
    },
    canSubPiAccess() {
      this.$acl.change(`${SUB_PI}`);
      return this.$acl.check('userAccess');
    },
    canScAccess() {
      this.$acl.change(`${STUDY_COORDINATOR}`);
      return this.$acl.check('userAccess');
    },
    canHomeHealthAccess() {
      this.$acl.change(`${HOME_HEALTH}`);
      return this.$acl.check('userAccess');
    },
    canCraAccess() {
      this.$acl.change(`${CRA}`);
      return this.$acl.check('userAccess');
    },
    clearFormName() {
      return this.formName ? this.formName.replace(/_/g, ' ') : '';
    },
    verified() {
      return this.verificationStatus === verificationStatus.VERIFIED;
    },
    approved() {
      return this.approvalStatus === approvalStatus.APPROVED;
    },
    verifyBtnText() {
      return this.verified ? 'Unverify' : 'Verify';
    },
    approvalBtnText() {
      return this.approved ? 'Unapprove' : 'Approve';
    },
  },
  methods: {
    openVerificationModal() {
      this.$refs.verificationModal.show();
    },
    openApprovalModal() {
      this.$refs.approvalModal.show();
    },
    verificationConfirmed({ verified }) {
      const status = verified ? verificationStatus.VERIFIED : verificationStatus.NOT_VERIFIED;
      this.$emit('verificationConfirmed', status);
    },
    approvalConfirmed({ approved }) {
      const status = approved ? approvalStatus.APPROVED : approvalStatus.NOT_APPROVED;
      this.$emit('approvalConfirmed', status);
    },
  },
};
</script>

<style scoped>
.APPROVED,
.VERIFIED {
  color: #48bf7a !important;
}
.NOT_VERIFIED,
.NOT_APPROVED {
  color: #f16559 !important;
}
.form-action-button {
  width: 100%;
  margin-top: 1px;
  margin-bottom: 1px;
}
dt {
  font-size: 10px;
  line-height: 1.6;
  text-transform: uppercase;
  color: #adaeb1;
}
dd {
  color: #666669;
  font-size: 14px;
  font-weight: 500;
}
.date {
  font-size: 12px;
  line-height: 1.67;
}
</style>
