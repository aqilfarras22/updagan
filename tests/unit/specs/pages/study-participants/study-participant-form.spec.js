import BootstrapVue from 'bootstrap-vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import studyParticipantsFormPage from '@/pages/study-participants/study-participant-form';
import { formFieldQueryStatus } from '@/constants/form';

const { it, describe, expect } = global;

describe('pages/study-participants/study-participant-details.vue', () => {
  const localVue = createLocalVue();
  localVue.use(BootstrapVue);
  const studyIdProp = '100';
  const formId = 111;

  const createMountedWrapper = testMocks => {
    const getQueryToolStudyFormResponse = {
      generalInfo: { name: 'name' },
      participant: { name: 'name' },
      sections: [
        {
          id: 1,
        },
        {
          id: 2,
        },
      ],
      dataQueries: [
        {
          id: 1,
        },
        {
          id: 2,
        },
      ],
    };

    const getQueryToolStudyFormHistoryResponse = [];

    const getQueryToolStudyFormStub = jest.fn().mockResolvedValue(getQueryToolStudyFormResponse);
    const getQueryToolStudyFormHistoryStub = jest
      .fn()
      .mockResolvedValue(getQueryToolStudyFormHistoryResponse);

    const routerGoStub = jest.fn();
    const toastrSuccessStub = jest.fn();
    const setQueryToolFormQueryStatusStub = jest.fn().mockResolvedValue({});
    const getQueryToolStudyFormHistoryVersionStub = jest.fn().mockResolvedValue({});
    const addQueryToolFormQueryCommentStub = jest.fn().mockResolvedValue({});
    const addQueryToolFieldQueryStub = jest.fn().mockResolvedValue({});
    const saveQueryToolStudyFormStub = jest.fn().mockResolvedValue({});
    const vueAclCheckStub = jest.fn().mockResolvedValue({});
    const vueAclChangeStub = jest.fn().mockResolvedValue({});
    return {
      wrapper: shallowMount(studyParticipantsFormPage, {
        localVue,
        propsData: {
          studyId: studyIdProp,
        },
        mocks: {
          $api: {
            getQueryToolStudyForm: getQueryToolStudyFormStub,
            getQueryToolStudyFormHistory: getQueryToolStudyFormHistoryStub,
            setQueryToolFormQueryStatus: setQueryToolFormQueryStatusStub,
            getQueryToolStudyFormHistoryVersion: getQueryToolStudyFormHistoryVersionStub,
            addQueryToolFormQueryComment: addQueryToolFormQueryCommentStub,
            addQueryToolFieldQuery: addQueryToolFieldQueryStub,
            saveQueryToolStudyForm: saveQueryToolStudyFormStub,
          },
          $toastr: {
            success: toastrSuccessStub,
          },
          $route: {
            params: {
              formId,
            },
          },
          $router: {
            go: routerGoStub,
          },
          $acl: {
            check: vueAclCheckStub,
            change: vueAclChangeStub,
          },
          ...testMocks,
        },
        stubs: {
          ParticipantInfo: { name: 'ParticipantInfo', render: h => h('b', 'ParticipantInfoCmp') },
          FormInfo: { name: 'FormInfo', render: h => h('b', 'FormInfoCmp') },
          FormHistoryList: { name: 'FormHistoryList', render: h => h('b', 'FormHistoryListCmp') },
          FormHistoryVersion: {
            name: 'FormHistoryVersion',
            render: h => h('b', 'FormHistoryVersionCmp'),
          },
          FormFieldsList: { name: 'FormFieldsList', render: h => h('b', 'FormFieldsListCmp') },
          FormFieldsEditList: {
            name: 'FormFieldsEditList',
            render: h => h('b', 'FormFieldsEditListCmp'),
          },
          FormQueries: { name: 'FormQueries', render: h => h('b', 'FormQueriesCmp') },
          FormQueriesHeader: {
            name: 'FormQueriesHeader',
            render: h => h('b', 'FormQueriesHeaderCmp'),
          },
          ButtonCmp: { name: 'ButtonCmp', render: h => h('b', 'ButtonCmp') },
        },
      }),
      setQueryToolFormQueryStatusStub,
      addQueryToolFormQueryCommentStub,
      saveQueryToolStudyFormStub,
      addQueryToolFieldQueryStub,
      getQueryToolStudyFormHistoryVersionStub,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormResponse,
      getQueryToolStudyFormHistoryStub,
      getQueryToolStudyFormHistoryResponse,
      routerGoStub,
      toastrSuccessStub,
    };
  };

  it('should create instance', () => {
    const { wrapper } = createMountedWrapper();
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it('should have described child components', () => {
    expect(studyParticipantsFormPage.components.ParticipantInfo).toBeTruthy();
    expect(studyParticipantsFormPage.components.FormInfo).toBeTruthy();
    expect(studyParticipantsFormPage.components.FormHistoryList).toBeTruthy();
    expect(studyParticipantsFormPage.components.FormHistoryVersion).toBeTruthy();
    expect(studyParticipantsFormPage.components.FormFieldsList).toBeTruthy();
    expect(studyParticipantsFormPage.components.FormFieldsEditList).toBeTruthy();
    expect(studyParticipantsFormPage.components.FormQueries).toBeTruthy();
    expect(studyParticipantsFormPage.components.FormQueriesHeader).toBeTruthy();
    expect(studyParticipantsFormPage.components.ButtonCmp).toBeTruthy();
  });

  it('should be able to get required props', () => {
    const { wrapper } = createMountedWrapper();
    const props = {
      studyId: '1',
      studyData: { data: 'data' },
    };
    wrapper.setProps(props);
    expect(wrapper.vm.studyId).toEqual(props.studyId);
    expect(wrapper.vm.studyData).toEqual(props.studyData);
  });

  it('should have all required initial data', () => {
    const { wrapper } = createMountedWrapper();
    // initial data
    expect(wrapper.vm.tabIndex).toBe(0);
    expect(wrapper.vm.participantInfo).toEqual(null);
    expect(wrapper.vm.formHistory).toEqual(null);
    expect(wrapper.vm.newQuery).toEqual(null);
    expect(wrapper.vm.formHistoryVersion).toEqual(null);
    expect(wrapper.vm.isLoading).toBe(true);
    expect(wrapper.vm.edit).toBe(false);
    expect(wrapper.vm.filterFieldId).toEqual(null);
    expect(wrapper.vm.generalInfo).toEqual(null);
    expect(wrapper.vm.participant).toEqual(null);
    expect(wrapper.vm.sections).toEqual([]);
    expect(wrapper.vm.queries).toEqual([]);
  });

  it('should return computed property: fId', () => {
    const { wrapper } = createMountedWrapper();
    expect(wrapper.vm.fId).toBe(formId);
  });

  it('should return computed property: fieldsWithQueries', () => {
    const { wrapper } = createMountedWrapper();
    expect(wrapper.vm.fieldsWithQueries).toEqual([]);
    const queries = [
      {
        fieldId: 1,
        status: formFieldQueryStatus.OPEN,
      },
      {
        fieldId: 2,
        status: formFieldQueryStatus.CLOSE,
      },
      {
        fieldId: 3,
        status: formFieldQueryStatus.OPEN,
      },
    ];
    wrapper.setData({
      queries,
    });
    expect(wrapper.vm.fieldsWithQueries).toEqual([1, 3]);
  });

  it('should have method which request all required data: getData', async () => {
    const {
      wrapper,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormResponse,
      getQueryToolStudyFormHistoryStub,
      getQueryToolStudyFormHistoryResponse,
    } = createMountedWrapper();
    wrapper.setData({
      isLoading: true,
    });
    // method is called on mount
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormStub).toBeCalledWith({
      fId: formId,
    });
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalledWith({
      fId: formId,
    });

    // getData should change state
    await getQueryToolStudyFormStub();
    await getQueryToolStudyFormHistoryStub();

    expect(wrapper.vm.isLoading).toBe(false);
    expect(wrapper.vm.generalInfo).toEqual(getQueryToolStudyFormResponse.generalInfo);
    expect(wrapper.vm.sections).toEqual(getQueryToolStudyFormResponse.sections);
    expect(wrapper.vm.queries).toEqual(getQueryToolStudyFormResponse.dataQueries);
    expect(wrapper.vm.participantInfo).toEqual(getQueryToolStudyFormResponse.participant);
    expect(wrapper.vm.formHistory).toEqual(getQueryToolStudyFormHistoryResponse);
  });

  it('should have method which reopen query: onQueryReopened', async () => {
    const {
      wrapper,
      setQueryToolFormQueryStatusStub,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormHistoryStub,
      toastrSuccessStub,
    } = createMountedWrapper();
    const query = {
      queryId: 1,
    };
    wrapper.vm.onQueryReopened(query);

    expect(setQueryToolFormQueryStatusStub).toBeCalled();
    expect(setQueryToolFormQueryStatusStub).toBeCalledWith({
      qId: query.queryId,
      patchId: 'open',
    });

    await setQueryToolFormQueryStatusStub();

    // should notify
    expect(toastrSuccessStub).toBeCalled();
    expect(toastrSuccessStub).toBeCalledWith('Query reopened');

    // should update page data
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
  });

  it('should have method which close query: onQueryClosed', async () => {
    const {
      wrapper,
      setQueryToolFormQueryStatusStub,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormHistoryStub,
      toastrSuccessStub,
    } = createMountedWrapper();
    const query = {
      queryId: 1,
    };
    wrapper.vm.onQueryClosed(query);

    expect(setQueryToolFormQueryStatusStub).toBeCalled();
    expect(setQueryToolFormQueryStatusStub).toBeCalledWith({
      qId: query.queryId,
      patchId: 'close',
    });

    await setQueryToolFormQueryStatusStub();

    // should notify
    expect(toastrSuccessStub).toBeCalled();
    expect(toastrSuccessStub).toBeCalledWith('Query closed');

    // should update page data
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
  });

  it('should have method which get form history version: viewHistoryForm', async () => {
    const { wrapper, getQueryToolStudyFormHistoryVersionStub } = createMountedWrapper();
    expect(wrapper.vm.formHistoryVersion).toEqual(null);
    wrapper.vm.viewHistoryForm({ versionId: 1 });

    await getQueryToolStudyFormHistoryVersionStub();

    expect(getQueryToolStudyFormHistoryVersionStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryVersionStub).toBeCalledWith({
      vId: 1,
      fId: formId,
    });
    expect(wrapper.vm.formHistoryVersion).toEqual({});
  });

  it('should have method which post a comment: createComment', async () => {
    const {
      wrapper,
      addQueryToolFormQueryCommentStub,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormHistoryStub,
    } = createMountedWrapper();
    wrapper.vm.createComment({ comment: 'comment' });

    await addQueryToolFormQueryCommentStub();

    expect(addQueryToolFormQueryCommentStub).toBeCalled();
    expect(addQueryToolFormQueryCommentStub).toBeCalledWith({ comment: 'comment' });

    // should update page data
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
  });

  it('should have method which save form: saveForm', async () => {
    const {
      wrapper,
      saveQueryToolStudyFormStub,
      getQueryToolStudyFormStub,
      toastrSuccessStub,
      getQueryToolStudyFormHistoryStub,
    } = createMountedWrapper();

    const form = {
      sections: [],
      primaryReason: 'reason',
    };
    wrapper.setData({
      edit: true,
    });
    wrapper.vm.saveForm(form);

    await saveQueryToolStudyFormStub();

    expect(saveQueryToolStudyFormStub).toBeCalled();
    expect(saveQueryToolStudyFormStub).toBeCalledWith({ fId: formId }, form);

    expect(wrapper.vm.edit).toBe(false);

    // should notify
    expect(toastrSuccessStub).toBeCalled();
    expect(toastrSuccessStub).toBeCalledWith('Edits saved');

    // should update page data
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
  });

  it('should have method which post a query: createQuery', async () => {
    const {
      wrapper,
      addQueryToolFieldQueryStub,
      getQueryToolStudyFormStub,
      toastrSuccessStub,
      getQueryToolStudyFormHistoryStub,
    } = createMountedWrapper();

    wrapper.vm.createQuery({ comment: 'comment' });

    await addQueryToolFieldQueryStub();

    expect(addQueryToolFieldQueryStub).toBeCalled();
    expect(addQueryToolFieldQueryStub).toBeCalledWith({ comment: 'comment' });

    // should notify
    expect(toastrSuccessStub).toBeCalled();
    expect(toastrSuccessStub).toBeCalledWith('Query created');

    // should update page data
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
  });

  it('should have method to handle verification confirmation: verificationConfirmed', () => {
    const {
      wrapper,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormHistoryStub,
    } = createMountedWrapper();
    wrapper.setData({
      generalInfo: { verificationStatus: false },
    });
    wrapper.vm.verificationConfirmed(true);
    expect(wrapper.vm.generalInfo.verificationStatus).toBe(true);
    wrapper.vm.verificationConfirmed(false);
    expect(wrapper.vm.generalInfo.verificationStatus).toBe(false);

    // should update page data
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
  });

  it('should have method to handle approval confirmation: approvalConfirmed', () => {
    const {
      wrapper,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormHistoryStub,
    } = createMountedWrapper();
    wrapper.setData({
      generalInfo: { approvalStatus: false },
    });
    wrapper.vm.approvalConfirmed(true);
    expect(wrapper.vm.generalInfo.approvalStatus).toBe(true);
    wrapper.vm.approvalConfirmed(false);
    expect(wrapper.vm.generalInfo.approvalStatus).toBe(false);

    // should update page data
    expect(getQueryToolStudyFormStub).toBeCalled();
    expect(getQueryToolStudyFormHistoryStub).toBeCalled();
  });

  it('should have method which cancel edit mode: cancelEdit', () => {
    const { wrapper } = createMountedWrapper();
    wrapper.setData({
      edit: true,
    });
    wrapper.vm.cancelEdit();
    expect(wrapper.vm.edit).toBe(false);
  });

  it('should have method which set field id to new query: openQuery', () => {
    const { wrapper } = createMountedWrapper();
    expect(wrapper.vm.newQuery).toEqual(null);
    wrapper.vm.openQuery(1);
    expect(wrapper.vm.newQuery).toBe(1);
  });
  it('should have method which reset history mode: historyBack', () => {
    const { wrapper } = createMountedWrapper();
    wrapper.setData({
      formHistoryVersion: 1,
    });
    wrapper.vm.historyBack();
    expect(wrapper.vm.formHistoryVersion).toEqual(null);
  });

  it('should have method which set filtered field: filterByField', () => {
    const { wrapper } = createMountedWrapper();
    expect(wrapper.vm.filterFieldId).toEqual(null);
    wrapper.vm.filterByField(1);
    expect(wrapper.vm.filterFieldId).toBe(1);
  });

  it('should have method which reset filtered field: resetFilter', () => {
    const { wrapper } = createMountedWrapper();
    expect(wrapper.vm.filterFieldId).toEqual(null);
    wrapper.vm.filterByField(1);
    expect(wrapper.vm.filterFieldId).toBe(1);
    wrapper.vm.resetFilter();
    expect(wrapper.vm.filterFieldId).toEqual(null);
  });

  it('should have method which route to previous page: goToPreviousScreen', () => {
    const { wrapper, routerGoStub } = createMountedWrapper();
    wrapper.vm.goToPreviousScreen();
    expect(routerGoStub).toBeCalled();
    expect(routerGoStub).toBeCalledWith(-1);
  });

  it('should has the expected html structure', async () => {
    const {
      wrapper,
      getQueryToolStudyFormStub,
      getQueryToolStudyFormHistoryStub,
    } = createMountedWrapper();
    await getQueryToolStudyFormStub();
    await getQueryToolStudyFormHistoryStub();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot();
  });
});
