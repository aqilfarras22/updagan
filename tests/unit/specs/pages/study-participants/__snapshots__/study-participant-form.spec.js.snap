// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/study-participants/study-participant-details.vue should has the expected html structure 1`] = `
<b-container-stub
  class="pt-30 mb-5 component"
  fluid="true"
  tag="div"
>
  <b-row-stub
    tag="div"
  >
    <b-col-stub
      cols="2"
      tag="div"
    >
      <b
        class="back-button"
        size="sm"
        variant="link"
      >
        ButtonCmp
      </b>
       
      <b
        name="name"
      >
        FormInfoCmp
      </b>
       
      <hr
        class="mt-30 mb-30"
      />
       
      <b
        participantdetails="[object Object]"
      >
        ParticipantInfoCmp
      </b>
    </b-col-stub>
     
    <b-col-stub
      class="pb-20"
      cols="10"
      tag="div"
    >
      <b-card-stub
        bodytag="div"
        class="border-0 shadow study-particip-form-block"
        footertag="div"
        headertag="div"
        nobody="true"
        subtitletag="h6"
        tag="div"
        titletag="h4"
      >
        <b-row-stub
          nogutters="true"
          tag="div"
        >
          <b-col-stub
            cols="7"
            tag="div"
            xl="8"
          >
            <b-tabs-stub
              tag="div"
              value="0"
            >
              <b-tab-stub
                buttonid=""
                href="#"
                tag="div"
                title="Latest Version"
              >
                <b-row-stub
                  tag="div"
                >
                  <b-col-stub
                    tag="div"
                  >
                    <div
                      class="float-right my-1"
                    >
                      <b
                        size="sm"
                        variant="outline-secondary"
                      >
                        ButtonCmp
                      </b>
                    </div>
                  </b-col-stub>
                </b-row-stub>
                 
                <b
                  fieldswithqueries=""
                  sections="[object Object],[object Object]"
                >
                  FormFieldsListCmp
                </b>
                 
                <!---->
              </b-tab-stub>
               
              <b-tab-stub
                buttonid=""
                href="#"
                tag="div"
                title="History"
              >
                <b
                  history=""
                >
                  FormHistoryListCmp
                </b>
              </b-tab-stub>
            </b-tabs-stub>
          </b-col-stub>
           
          <b-col-stub
            class="tabs-queries"
            cols="5"
            tag="div"
            xl="4"
          >
            <b
              formid="111"
              queries="[object Object],[object Object]"
            >
              FormQueriesCmp
            </b>
          </b-col-stub>
        </b-row-stub>
      </b-card-stub>
    </b-col-stub>
  </b-row-stub>
</b-container-stub>
`;
